File: problems-adapt/0401-0500/0436_transit-trip-log/solution.go
package main

type tripCheckin struct {
	stop string
	time int
}

type tripKey struct {
	from string
	to   string
}

type tripTotal struct {
	sum   int64
	count int64
}

type TransitLog struct {
	checkins map[int]tripCheckin
	totals   map[tripKey]tripTotal
}

func NewTransitLogTyped() *TransitLog {
	return &TransitLog{}
}

func (design *TransitLog) ensure() {
	if design.checkins == nil {
		design.checkins = make(map[int]tripCheckin)
		design.totals = make(map[tripKey]tripTotal)
	}
}

func (design *TransitLog) tapIn(id int, stop string, t int) {
	design.ensure()
	design.checkins[id] = tripCheckin{stop: stop, time: t}
}

func (design *TransitLog) tapOut(id int, stop string, t int) {
	design.ensure()
	start := design.checkins[id]
	delete(design.checkins, id)
	key := tripKey{from: start.stop, to: stop}
	bucket := design.totals[key]
	bucket.sum += int64(t - start.time)
	bucket.count++
	design.totals[key] = bucket
}

func (design *TransitLog) averageTrip(fromStop string, toStop string) float64 {
	bucket := design.totals[tripKey{from: fromStop, to: toStop}]
	return float64(bucket.sum) / float64(bucket.count)
}

File: problems-adapt/0401-0500/0436_transit-trip-log/solution.py
class TransitLog:
    def __init__(self) -> None:
        # trips in flight: id -> (station, t). Only the pending check-in
        # is needed until the customer checks out
        self.checkins: dict[int, tuple[str, int]] = {}
        # completed trips keyed by ordered station pair -> [total time, count]
        self.totals: dict[tuple[str, str], list[int]] = {}

    def tapIn(self, id: int, stop: str, t: int) -> None:
        self.checkins[id] = (stop, t)

    def tapOut(self, id: int, stop: str, t: int) -> None:
        start, started = self.checkins.pop(id)
        # the trip collapses to one duration folded into a per-pair sum+count;
        # consuming the check-in frees the id to travel again immediately
        bucket = self.totals.setdefault((start, stop), [0, 0])
        bucket[0] += t - started
        bucket[1] += 1

    def averageTrip(self, fromStop: str, toStop: str) -> float:
        # no durations are ever stored — just divide the running totals
        total, count = self.totals[(fromStop, toStop)]
        return total / count
